include/swift/AST/ASTContext.h

```diff
@@ -561,6 +562,15 @@ class ASTContext final {
   Type getBridgedToObjC(const DeclContext *dc, Type type,
                         Type *bridgedValueType = nullptr) const;
 
+  /// Get the Clang type corresponding to a Swift function type.
+  ///
+  /// \param resultTy The Swift result type.
+  /// \param paramsTy The Swift parameter type.
+  /// \param repr The Swift calling convention, which must be C-compatible.
+  const clang::Type *
+  getClangFunctionType(Type resultTy, ArrayRef<Type> paramsTy,
+                       FunctionTypeRepresentation repr);
```

Comment:
Can you make these arguments exactly match the normal inputs to `getFunctionType`, so that it takes an array of `AnyFunctionType::Param` and an `ExtInfo` (albeit one with either the wrong representation or (obviously) a missing Clang type)?

Reply:
> Can you make these arguments exactly match the normal inputs to getFunctionType, so that it takes an array of AnyFunctionType::Param and an ExtInfo

I'm guessing you meant `swift::FunctionType::get` here? Could you point out why this is beneficial?

The order of the arguments is written so as to match the Clang convention (result type first) instead of the Swift convention (params first). I suppose that is a bit questionable, since we are passing Swift types in...

> albeit one with either the wrong representation or (obviously) a missing Clang type

Both of these options are unappealing IMO because (a) the first one is just wrong and (b) the second one means we won't be able to add the invariant in the future that we always have a clang type. I want to ensure correctness-by-construction so that we have less confusing moments when debugging like "oh, this is in an incorrect/invalid state here but there's a good reason for that".

Reply:
I don't care about the exact order of arguments.

You should take an array of `Param`s instead of `Type`s because we very well might want to generate a different Clang type based on the parameter attributes.  It also looks like it'll generally be easier for the call sites.

The argument for ExtInfo is similar, in that we might well want to change the Clang function type based on Swift function-level attributes.  For example, we currently don't allow `@convention(c)` function types to be throwing, but if we did, it would certainly affect the exported Clang function type.  Experience suggests that the set of function attributes tends to grow over time.  I agree with your argument against allowing the construction of an ExtInfo with C representation and no type, but I think you can reasonably pass down an ExtInfo with a comment that the representation is meaningless and then pass down the real representation separately.

Reply:
Got it, thanks for explaining!